### ISSUE TITLE
Optimizer should handle Table Updates for indexed attributes

### ISSUE THREAD

**Issue by thepulkitagarwal** [cmu-db/terrier#592]
The optimizer should ensure that:
- If no indexed attributes are updated:
  - It updates the tuple slot.
  - It shouldn't modify any indexes.
- If any indexed attributes are updated:
  - It deletes the tuple slot from the table, and re-inserts to get a new tuple slot.
  - For all indexes, it deletes and then inserts.

**Comment by lmwnshn** [comment]
I can confirm that this issue continues to exist in today's version of Terrier.

```
terrier=# create table foo (a int primary key, b int);
CREATE TABLE
terrier=# insert into foo values (1,2);
INSERT 0 1
terrier=# insert into foo values (2,3);
INSERT 0 1
terrier=# select * from foo;
 a | b 
---+---
 1 | 2
 2 | 3
(2 rows)

terrier=# update foo set a=2 where b=2;
UPDATE 1
terrier=# select * from foo;
 a | b 
---+---
 2 | 2
 2 | 3
(2 rows)
```

The problem is,
1. Because the optimizer always has indexed updates set to false, we're not doing the delete-and-insert on indexes in codegen.
2. I think @tanujnay112 fixed it (to correctly identify indexed updates in some cases) at [0].
3. The old and new execution engine both blow up after the fix. The problem is that the update ProjectedRow does not contain the full row and therefore cannot be used for inserting back into the table. I am unsure what the right approach is / we should probably discuss this.
4. For now, I am re-breaking the optimizer so that we can continue porting. We won't be any more broken than the old engine anyway...

@17zhangw If you want to weigh in on the fix?

[0] https://github.com/lmwnshn/terrier/blob/pmenon_vpi/src/optimizer/plan_generator.cpp#L795

**Comment by lmwnshn** [comment]
Per conversation with @17zhangw, it works if the updated columns cover all attributes in the table,

```
terrier=# create table foo (a int primary key, b int);
CREATE TABLE
terrier=# insert into foo values (1,2);
INSERT 0 1
terrier=# update foo set a=2,b=3 where b=2;
UPDATE 2
terrier=# select * from foo;
 a | b 
---+---
 2 | 3
(1 row)
```

**Comment by lmwnshn** [comment]
Fixed by @tanujnay112 in [this commit](https://github.com/cmu-db/terrier/pull/994/commits/bdc1008ef0299e246808f3e5d7db7711cb8e3646) as part of TPL2.

**Comment by lmwnshn** [comment]
Fixed by #994.
